graphql_compiler/schema_transformation/utils.py

```diff
@@ -0,0 +1,135 @@
+        """Create a visitor for checking query type field names.
+
+        Args:
+            query_type: str, name of the query type (e.g. RootSchemaQuery)
+        """
+        self.query_type = query_type
+        self.in_query_type = False
+
+    def enter_ObjectTypeDefinition(self, node, *args):
+        """If the node's name matches the query type, record that we entered the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = True
+
+    def leave_ObjectTypeDefinition(self, node, *args):
+        """If the node's name matches the query type, record that we left the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = False
+
+    def enter_FieldDefinition(self, node, *args):
+        """If inside the query type, check that the field and queried type names match.
+
+        Raises:
+            SchemaStructureError if the field name is not identical to the name of the type
+            that it queries
+        """
+        if self.in_query_type:
+            field_name = node.name.value
+            type_node = node.type
+            # NamedType node may be wrapped in several layers of NonNullType or ListType
+            while not isinstance(type_node, NamedType):
+                type_node = type_node.type
+            queried_type_name = type_node.name.value
+            if field_name != queried_type_name:
+                raise SchemaStructureError(
+                    u'Query type\'s field name "{}" does not match corresponding queried type '
+                    'name "{}"'.format(field_name, queried_type_name)
+                )
+
+
+def _check_query_type_fields_name_match(ast, query_type):
+    """Check every query type field's name is identical to the type it queries.
+
+    Args:
+        ast: Document representing a schema
+        query_type: str, name of the query type
+
+    Raises:
+        SchemaStructureError if any query type field name is not identical to the name of the
+        type that it queries
+    """
+    visitor = CheckQueryTypeFieldsNameMatchVisitor(query_type)
+    visit(ast, visitor)
+
+
+def check_ast_schema_is_valid(ast, schema):
+    """Check the schema satisfies structural requirements for rename and merge.
+
+    In particular, check that the schema contains no mutations, no subscriptions, and all query
+    type field names match the types they query.
+
+    Args:
+        ast: Document, representing a schema
+        schema: GraphQLSchema, representing the same schema as ast
+
+    Raises:
+        SchemaStructureError if the schema contains mutations, contains subscriptions, or some
+        query type field name does not match the type it queries.
+    """
+    if schema.get_mutation_type() is not None:
+        raise SchemaStructureError(u'Schema contains mutations.')
+    if schema.get_subscription_type() is not None:
+        raise SchemaStructureError(u'Schema contains subscriptions.')
```

Comment:
I would add to the error messages something like: `Renaming schemas that contain X is currently not supported.` for the appropriate values of `X`.